db/db_impl.h

```diff
@@ -1565,6 +1579,9 @@ class DBImpl : public DB {
     return Env::WLTH_SHORT;
   }
 
+  std::map<uint64_t, std::map<std::string, std::string>>
+  FindStatsBetween(uint64_t start_time, uint64_t end_time);
```

Comment:
I think this API needs more people to review. I am concerned to it. It assumes that all results need to be copied to memory. This makes memory budget hard to control. Yes we can add one more argument about maximum records to report, but this makes the function harder to understand. Another possible way is to make it an iterator interface, so that we only need to page in the records people are currently reading.
It's worth asking MyRocks people about how they plan to turn this interface into a virtual SQL table and allow arbitrary SQL queries, and we should enable them to do it through our API.

Reply:
Right now we have discussed two general approaches to storing and retrieving stats history. 
* in-memory: implemented in this PR. Basic idea is to periodically store stats into a map and query through a new API. 
  * pros: no backward compatibility issue
  * cons: increased memory consumption/limited length of stats history (unable to query history before the last DB restart)
* persist to disk: implemented in https://github.com/facebook/rocksdb/pull/4535. Basic idea is to implicitly create a new column family to store stats history, and use iterator interface to query
  * pros: limited memory cost, supports queries over lifetime of the DB
  * cons: additional downgrade operational cost (to manage the new column family), complication with managing additional column family

The original plan was to first roll out in-memory solution to get user engagement with the new API, and then roll out persist to disk when user are comfortable with the potential operational cost of downgrading or find a better solution for smooth downgrade.
@yoshinorim do you mind taking a look and let us know your preference? especially with MyRocks integration?

Reply:
I think the roll out plan makes sense. From high level point of view, memory consumption is probably the biggest concern.
- Capping memory usage if Stats History consumes too much (capping the configurable number of entries or memory size. The latter is easier to tune for users)
- Reporting rough memory size used by this feature

Reply:
@miasantreble  I am lost in your comment. I've never suggested any "new API". Both in-memory and persistent-to-disk solution should use the same API. The benefit of having the in-memory version at all is that people can adapt to it, and then later naturally migrate to the persistence version without any API change.

Reply:
@siying thanks for pointing out the ambiguity, I've updated the comment above

Reply:
If holding the stats in-memory is a temporary feature I'd be careful about adding an option especially for it.

In the persistent version of the stats, will there be any API to control how much storage is used for stats? Or how much time we retain stats? If so, perhaps we can use that same API to control memory usage when we have the in-memory implementation.

Reply:
@ajkr maybe we can add another API `StartCollectStatsHistory(GetStatsOptions&)` to start the background stats collection threads where we can specify the memory limit in `GetStatsOptions`, and to retrieve most recent stats history, still use `GetStatsHistory`. Basically we can put all new options in `GetStatsOptions` to avoid polluting the general options space. How does that sound?

Reply:
@siying I gave some thought about the iterator interface. My main concern is that the stats history data structure (current a std::map) will be constantly updated (inserted and erased), and iterator interface will need to either 1) delay purging expired data to ensure iterator will have a consistent view of the data, or 2) create possible race condition with iterator reads and data purging. Yoshi pointed to me a similar case in MyRocks where we need to create virtual table using RocksDB data structure: https://github.com/facebook/mysql-5.6/blob/60ab18528d63709cc82364d5b9b81ecbfe8d86e8/storage/rocksdb/rdb_i_s.cc#L1148
MyRocks calls GetPropertiesOfAllTables which takes a pointer to a `std::unordered_map<std::string, std::shared_ptr<const TableProperties>>`. We can do similar things here to save some parameter copying cost

Reply:
> @siying I gave some thought about the iterator interface. My main concern is that the stats history data structure (current a std::map) will be constantly updated (inserted and erased), and iterator interface will need to either 1) delay purging expired data to ensure iterator will have a consistent view of the data, or 2) create possible race condition with iterator reads and data purging. Yoshi pointed to me a similar case in MyRocks where we need to create virtual table using RocksDB data structure: https://github.com/facebook/mysql-5.6/blob/60ab18528d63709cc82364d5b9b81ecbfe8d86e8/storage/rocksdb/rdb_i_s.cc#L1148

Is the in-memory history stats data a permanent feature? If not, I think we could perhaps prioritize API design over consideration of specific (in-memory) data structures, e.g. `std::map`. Once we have an interface, we can find a in-memory data structure that meets the requirement of the API.

Reply:
@miasantreble Let's put the in-memory case aside for just a moment. If we only consider the persistent case, what kind of function will you use?